lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,778 @@
+import re
+import selinux
+import shutil
+from typing import Dict, List
+
+from pbench.server import PbenchServerConfig
+from pbench.server.database.models.datasets import Dataset
+
+
+class FiletreeError(Exception):
+    """
+    Base class for exceptions raised from this code.
+    """
+
+    def __str__(self) -> str:
+        return "Generic file tree exception"
+
+
+class BadFilename(FiletreeError):
+    """
+    A bad path is given for a tarball.
+    """
+
+    def __init__(self, path: str):
+        self.path = str(path)
+
+    def __str__(self) -> str:
+        return f"The file path {self.path} is not a tarball"
+
+
+class DatasetNotFound(FiletreeError):
+    """
+    The on-disk representation of a dataset (tarball and MD5 companion) were
+    not found in the ARCHIVE tree.
+    """
+
+    def __init__(self, dataset: str):
+        self.dataset = dataset
+
+    def __str__(self) -> str:
+        return f"The dataset named {self.dataset!r} is not present in the file tree"
+
+
+class DuplicateDataset(FiletreeError):
+    """
+    A duplicate dataset name was detected.
+    """
+
+    def __init__(self, dataset: str):
+        self.dataset = dataset
+
+    def __str__(self) -> str:
+        return f"A dataset named {self.dataset!r} is already present in the file tree"
+
+
+class ControllerNotFound(FiletreeError):
+    """
+    A specified controller name does not exist in the ARCHIVE tree.
+    """
+
+    def __init__(self, controller: str):
+        self.controller = controller
+
+    def __str__(self) -> str:
+        return (
+            f"The controller name {self.controller!r} is not present in the file tree"
+        )
+
+
+class Tarball:
+    """
+    This class corresponds to the physical representation of a Dataset: the
+    tarball, the MD5 file, and the unpacked data.
+
+    It provides discovery and management methods related to a specific
+    dataset.
+    """
+
+    @staticmethod
+    def stem(path: Path) -> str:
+        """
+        The Path.stem() removes a single suffix, so our standard "a.tar.xz"
+        returns "a.tar" instead of "a". We could double-stem, but instead
+        this just checks for the expected 7 character suffix and strips it.
+
+        If the path does not end in ".tar.xz" then the full path.name is
+        returned.
+
+        Args:
+            path: A file path that might be a Pbench tarball
+
+        Returns:
+            The stripped "stem" of the dataset
+        """
+        name = path.name
+        return name[:-7] if name[-7:] == ".tar.xz" else name
+
+    def __init__(self, path: Path, controller: "Controller"):
+        """
+        Construct a `Tarball` object instance representing a tarball found on
+        disk.
+
+        Args:
+            path: The file path to a discovered tarball (.tar.xz file) in the
+                configured ARCHIVE directory for a controller.
+            controller: The associated Controller object
+        """
+        self.name = Tarball.stem(path)
```

Comment:
I prefer using `self` for references from within the class because they are unaffected by renaming the class or by having the attribute hoisted into a superclass.  (Obviously, references from outside of the class don't have the luxury of using `self`.)